File: src/quest_editor/scripting/data_flow_analysis/ValueSet.ts
/**
 * Represents a sorted set of integers.
 */
export class ValueSet {
    /**
     * Closed intervals [start, end].
     */
    private intervals: { start: number; end: number }[] = [];

    size(): number {
        return this.intervals.reduce((acc, i) => acc + i.end - i.start + 1, 0);
    }

    get(i: number): number | undefined {
        for (const { start, end } of this.intervals) {
            const size = end - start + 1;

            if (i < size) {
                return start + i;
            } else {
                i -= size;
            }
        }

        return undefined;
    }

    min(): number | undefined {
        return this.intervals.length ? this.intervals[0].start : undefined;
    }

    max(): number | undefined {
        return this.intervals.length ? this.intervals[this.intervals.length - 1].end : undefined;
    }

    has(value: number): boolean {
        for (const int of this.intervals) {
            if (int.start <= value && value <= int.end) {
                return true;
            }
        }

        return false;
    }

    /**
     * Sets this ValueSet to the given integer.
     *
     * @param value integer value
     */
    set_value(value: number): ValueSet {
        this.intervals = [{ start: value, end: value }];
        return this;
    }

    /**
     * Sets this ValueSet to the values in the given interval.
     *
     * @param start lower bound, inclusive
     * @param end upper bound, inclusive
     */
    set_interval(start: number, end: number): ValueSet {
        if (end < start)
            throw new Error(
                `Interval upper bound should be greater than or equal to lower bound, got [${start}, ${end}].`,
            );

        this.intervals = [{ start, end }];
        return this;
    }

    /**
     * Doesn't take into account interger overflow.
     */
    scalar_add(s: number): ValueSet {
        for (const int of this.intervals) {
            int.start += s;
            int.end += s;
        }

        return this;
    }

    /**
     * Doesn't take into account interger overflow.
     */
    scalar_sub(s: number): ValueSet {
        return this.scalar_add(-s);
    }

    /**
     * Doesn't take into account interger overflow.
     */
    scalar_mul(s: number): ValueSet {
        for (const int of this.intervals) {
            int.start *= s;
            int.end *= s;
        }

        return this;
    }

    /**
     * Integer division.
     * Doesn't take into account interger overflow.
     */
    scalar_div(s: number): ValueSet {
        for (const int of this.intervals) {
            int.start = Math.floor(int.start / s);
            int.end = Math.floor(int.end / s);
        }

        return this;
    }

    union(other: ValueSet): ValueSet {
        let i = 0;

        outer: for (const b of other.intervals) {
            while (i < this.intervals.length) {
                const a = this.intervals[i];

                if (b.end < a.start - 1) {
                    this.intervals.splice(i, 0, b);
                    i++;
                    continue outer;
                } else if (b.start <= a.end + 1) {
                    a.start = Math.min(a.start, b.start);

                    let j = i;

                    while (j < this.intervals.length) {
                        if (b.end > this.intervals[j].start) {
                            a.end = this.intervals[j].end;
                            j++;
                        } else {
                            break;
                        }
                    }

                    this.intervals.splice(i + 1, j - i - 1);
                    a.end = Math.max(a.end, b.end);
                    i++;
                    continue outer;
                } else {
                    i++;
                }
            }

            this.intervals.push(b);
        }

        return this;
    }

    to_array(): number[] {
        let array: number[] = [];

        for (const { start, end } of this.intervals) {
            for (let i = start; i <= end; i++) {
                array.push(i);
            }
        }

        return array;
    }

    [Symbol.iterator](): Iterator<number> {
        const vs = this;
        let int_i = 0;
        let value = NaN;

        return {
            next(): IteratorResult<number> {
                let done = true;

                if (int_i < vs.intervals.length) {
                    if (isNaN(value)) {
                        value = vs.intervals[int_i].start;
                        done = false;
                    } else if (value > vs.intervals[int_i].end) {
                        int_i++;

                        if (int_i < vs.intervals.length) {
                            value = vs.intervals[int_i].start;
                            done = false;
                        }
                    } else {
                        done = false;
                    }
                }

                return { done, value: value++ };
            },
        };
    }
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/assembly/dataFlowAnalysis/ValueSet.kt
package world.phantasmal.lib.assembly.dataFlowAnalysis

import kotlin.math.max
import kotlin.math.min

/**
 * Represents a sorted set of integers.
 */
class ValueSet : Iterable<Int> {
    private val intervals: MutableList<Interval> = mutableListOf()

    val size: Int
        get() =
            intervals.fold(0) { acc, i -> acc + i.end - i.start + 1 }

    operator fun get(i: Int): Int? {
        var idx = i

        for ((start, end) in intervals) {
            val size = end - start + 1

            if (idx < size) {
                return start + idx
            } else {
                idx -= size
            }
        }

        return null
    }

    fun isEmpty(): Boolean =
        intervals.isEmpty()

    fun minOrNull(): Int? =
        intervals.firstOrNull()?.start

    fun maxOrNull(): Int? =
        intervals.lastOrNull()?.end

    operator fun contains(value: Int): Boolean {
        for (int in intervals) {
            if (value in int) {
                return true
            }
        }

        return false
    }

    /**
     * Sets this ValueSet to the given integer.
     */
    fun setValue(value: Int): ValueSet {
        intervals.clear()
        intervals.add(Interval(value, value))
        return this
    }

    /**
     * Sets this ValueSet to the values in the given interval.
     *
     * @param start lower bound, inclusive
     * @param end upper bound, inclusive
     */
    fun setInterval(start: Int, end: Int): ValueSet {
        require(end >= start) {
            "Interval upper bound should be greater than or equal to lower bound, got [${start}, ${end}]."
        }

        intervals.clear()
        intervals.add(Interval(start, end))
        return this
    }

    /**
     * Doesn't take into account integer overflow.
     */
    fun scalarAdd(s: Int): ValueSet {
        for (int in intervals) {
            int.start += s
            int.end += s
        }

        return this
    }

    /**
     * Doesn't take into account integer overflow.
     */
    fun scalarSub(s: Int): ValueSet {
        return scalarAdd(-s)
    }

    /**
     * Doesn't take into account integer overflow.
     */
    fun scalarMul(s: Int): ValueSet {
        for (int in intervals) {
            int.start *= s
            int.end *= s
        }

        return this
    }

    /**
     * Integer division.
     */
    fun scalarDiv(s: Int): ValueSet {
        for (int in intervals) {
            int.start = int.start / s
            int.end = int.end / s
        }

        return this
    }

    fun union(other: ValueSet): ValueSet {
        var i = 0

        outer@ for (b in other.intervals) {
            while (i < intervals.size) {
                val a = intervals[i]

                if (b.end < a.start - 1) {
                    // b lies entirely before a, insert it right before a.
                    intervals.add(i, b.copy())
                    i++
                    continue@outer
                } else if (b.start <= a.end + 1) {
                    // a and b overlap or form a continuous interval (e.g. [1, 2] and [3, 4]).
                    a.start = min(a.start, b.start)

                    // Merge all intervals that overlap with b.
                    val j = i + 1

                    while (j < intervals.size) {
                        if (b.end >= intervals[j].start - 1) {
                            a.end = intervals[j].end
                            intervals.removeAt(j)
                        } else {
                            break
                        }
                    }

                    a.end = max(a.end, b.end)
                    i++
                    continue@outer
                } else {
                    // b lies entirely after a, check next a.
                    i++
                }
            }

            // b lies after every a, add it to the end of our intervals.
            intervals.add(b.copy())
        }

        return this
    }

    override fun iterator(): Iterator<Int> =
        object : Iterator<Int> {
            private var intIdx = 0
            private var nextValue: Int? = minOrNull()

            override fun hasNext(): Boolean =
                nextValue != null

            override fun next(): Int {
                val v = nextValue ?: throw NoSuchElementException()

                nextValue =
                    if (v < intervals[intIdx].end) {
                        v + 1
                    } else {
                        intIdx++

                        if (intIdx < intervals.size) {
                            intervals[intIdx].start
                        } else {
                            null
                        }
                    }

                return v
            }
        }
}

/**
 * Closed interval [start, end].
 */
private data class Interval(var start: Int, var end: Int) {
    operator fun contains(value: Int): Boolean =
        value in start..end
}
